### PR TITLE
fix: handle shutdown gracefully

### PR DESCRIPTION
**Pull request by kurtisvg**
Adds logic to make the server shutdown gracefully, including better respecting cancelled contexts and providing up to 10 seconds to finish current connections.

### CODE REVIEW

File: cmd/root.go [hunk 1]
```diff
@@ -132,6 +135,26 @@ func run(cmd *Command) error {
...
 	defer cancel()
 
+	// watch for sigterm / sigint signals
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGTERM, syscall.SIGINT)
+	go func() {
+		var s os.Signal
+		select {
+		case <-ctx.Done():
+			// this should only happen when the context supplied in tests in canceled
```
Comment (duwenxin99) [inline cmd/root.go]:
```suggestion
			// this should only happen when the context supplied in tests is canceled
```

File: internal/server/server.go [hunk 2]
```diff
@@ -26,16 +26,18 @@ import (
...
 )
 
 // Server contains info for running an instance of Toolbox. Should be instantiated with NewServer().
 type Server struct {
-	conf   ServerConfig
-	root   chi.Router
-	logger logLib.Logger
+	version string
+	srv     *http.Server
+	l       net.Listener
```
Comment (duwenxin99) [inline internal/server/server.go]: Using `l` as listener here but as logger in NewServer(l log.Logger) is less readable. Maybe `listener` instead?

File: internal/server/server.go [hunk 3]
```diff
@@ -131,42 +145,47 @@ func NewServer(ctx context.Context, cfg ServerConfig, log logLib.Logger) (*Serve
...
 	}
-	log.InfoContext(ctx, fmt.Sprintf("Initialized %d toolsets.", len(toolsetsMap)))
+	l.InfoContext(ctx, fmt.Sprintf("Initialized %d toolsets.", len(toolsetsMap)))
+
+	addr := net.JoinHostPort(cfg.Address, strconv.Itoa(cfg.Port))
+	srv := &http.Server{Addr: addr, Handler: r}
 
-	s := &Server{
-		conf:        cfg,
+	s = &Server{
```
Comment (duwenxin99) [inline internal/server/server.go]: I think we need to set the `version`  here to `cfg.Version` otherwise it's lost

File: cmd/root.go [hunk 4]
```diff
@@ -132,6 +135,26 @@ func run(cmd *Command) error {
 	ctx, cancel := context.WithCancel(cmd.Context())
 	defer cancel()
 
+	// watch for sigterm / sigint signals
+	signals := make(chan os.Signal, 1)
+	signal.Notify(signals, syscall.SIGTERM, syscall.SIGINT)
+	go func() {
```
Comment (duwenxin99) [inline cmd/root.go:141]: QQ: why do we need to watch for interrupt signals before the server even gets started?

Comment (kurtisvg) [reply cmd/root.go:141]: We want to start watching as soon as the command starts to be sure that if something takes a long time while the command runs, it doesn't hold up the sigterm signal.

File: internal/server/server.go (hunk 3 above)
Comment (kurtisvg) [reply internal/server/server.go]: Good catch!

Comment (kurtisvg) [reply internal/server/server.go]: Added a test for this as well

**Pull request merged**